File: javascript/src/protection-proxy/protection-proxy.js
/** Checks the caller's role before forwarding to the subject. */
export class ProtectionProxy {
  #subject;
  #role;

  constructor(subject, role) {
    this.#subject = subject;
    this.#role = role;
  }

  request() {
    return this.#role === "admin" ? this.#subject.request() : "access denied by ProtectionProxy";
  }
}

File: python/src/patterns/protectionproxy/protection_proxy.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .subject import Subject


class ProtectionProxy:
    """Checks the caller's role before forwarding to the subject."""

    def __init__(self, subject: Subject, role: str) -> None:
        self._subject = subject
        self._role = role

    def request(self) -> str:
        if self._role == "admin":
            return self._subject.request()
        return "access denied by ProtectionProxy"
